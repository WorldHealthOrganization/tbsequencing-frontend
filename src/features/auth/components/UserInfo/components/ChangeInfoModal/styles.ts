import { css } from '@emotion/react';

export const buttonsContainer = css({
  marginTop: 32,
  justifyContent: 'flex-end',
  display: 'flex',
});

export const fieldWrapper = css({
  marginTop: 16,
});

export const changePasswordBtn = css({
  marginLeft: 16,
  justifyContent: 'flex-end',
});

export const fieldsContainer = css({
  marginTop: 32,
  marginBottom: 32,
  '.MuiInputLabel-root': {
    transform: 'translate(0, -1.5px) scale(0.75)',
    paddingLeft: '13px',
  },
});

export const baseModalSx = {
  overflowY: 'auto',
  height: '100%',
  display: 'block',
};
