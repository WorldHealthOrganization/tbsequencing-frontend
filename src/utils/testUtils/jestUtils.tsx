import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
