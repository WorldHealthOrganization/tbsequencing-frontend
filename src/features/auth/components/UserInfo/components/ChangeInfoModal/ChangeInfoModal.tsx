/** @jsxImportSource @emotion/react */
import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup/dist/yup';
import * as yup from 'yup';
import { toast } from 'react-toastify';
import * as styles from './styles';
import {
  validateEmail, validatePhone,
  validateTextField,
} from '../../../../../../utils/validators';
import H1 from '../../../../../../components/typography/H1';
import { AppButton } from '../../../../../../components/AppButton/AppButton';
import BaseModal from '../../../../../../components/BaseModal';
import { useChangeUserInfoMutation } from '../../../../../../services/indentityApi/identityApi';
import { IChangeUserInfoRequest } from '../../../../../../services/indentityApi/models';
import { useHttpError } from '../../../../../../hooks/useHttpError';
import { IHttpError } from '../../../../../../models/HttpError';

interface IInputs {
  firstName: string;
  lastName: string;
  institutionHeadName: string;
  institutionName: string;
  institutionPhone: string;
  institutionHeadEmail: string;
  institutionCountry: string;
}

interface Props {
  isOpened: boolean;
  handleClose: () => void;
}

const formSchema = yup.object().shape({
  firstName: validateTextField('First name is mandatory'),
  lastName: validateTextField('Last name is mandatory'),
  institutionHeadName: yup.string().required('Head of Lab is mandatory'),
  institutionName: yup.string().required('Institution is mandatory'),
  institutionPhone: validatePhone('Phone number is mandatory'),
  institutionHeadEmail: validateEmail('Email of Lab is mandatory'),
  institutionCountry: yup.string().required('Institution country is mandatory'),
});

export const ChangeInfoModal = ({
  isOpened, handleClose,
}: Props) => {
  const [changeUserInfo, { error, data }] = useChangeUserInfoMutation();
  const {
    formState: { isValid }, handleSubmit, setError, reset,
  } = useForm<IInputs>({
    resolver: yupResolver(formSchema),
    mode: 'onChange',
  });

  useHttpError<IInputs>(
    error as IHttpError,
    setError,
    'firstName',
  );

  useEffect(() => {
    if (data) {
      toast.success('User information was successfully changed');
      handleClose();
    }
  }, [data]);

  const onSubmit = (formData: IInputs) => {
    const requestParams: IChangeUserInfoRequest = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      profile: {
        institutionCountry: formData.institutionCountry,
        institutionName: formData.institutionName,
        institutionHeadEmail: formData.institutionHeadEmail,
        institutionPhone: formData.institutionPhone,
        institutionHeadName: formData.institutionHeadName,
      },
    };
    changeUserInfo(requestParams);
  };

  const onCancel = () => {
    handleClose();
    reset();
  };

  return (
    <BaseModal modalStyle={styles.baseModalSx} open={isOpened}>
      <div>
        <H1>Change info</H1>
        <div css={styles.buttonsContainer}>
          <AppButton onClick={onCancel} variant="outlined" size="large">Cancel</AppButton>
          <AppButton
            style={styles.changePasswordBtn}
            onClick={handleSubmit(onSubmit)}
            disabled={!isValid}
            variant="contained"
            size="large"
          >
            Save
          </AppButton>
        </div>
      </div>
    </BaseModal>
  );
};

export default ChangeInfoModal;
