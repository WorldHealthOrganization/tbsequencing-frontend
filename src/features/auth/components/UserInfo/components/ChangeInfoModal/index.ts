import ChangeInfoModal from './ChangeInfoModal';

export default ChangeInfoModal;
