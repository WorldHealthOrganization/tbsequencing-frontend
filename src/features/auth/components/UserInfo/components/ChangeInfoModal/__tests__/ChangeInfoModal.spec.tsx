import React from 'react';
import { render, getTreeJSONWithRedux } from '../../../../../../../utils/testUtils/jestUtils';
import ChangeInfoModal from '../ChangeInfoModal';

describe('ChangeInfoModal component test', () => {
  test('renders ChangeInfoModal component without error', () => {
    render(
      <ChangeInfoModal />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <ChangeInfoModal />,
    );

    expect(tree).toMatchSnapshot();
  });
});
