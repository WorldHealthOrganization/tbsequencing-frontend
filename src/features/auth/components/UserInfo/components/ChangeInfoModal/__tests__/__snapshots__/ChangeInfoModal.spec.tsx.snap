// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeInfoModal component test Should render correctly 1`] = `null`;
